sql/table.cc

```diff
@@ -8532,8 +8532,10 @@ int TABLE::update_generated_fields()
   {
     next_number_field= found_next_number_field;
     res= found_next_number_field->set_default();
-    if (likely(!res))
-      res= file->update_auto_increment();
+    if (likely(!res)) {
+      file->info(HA_STATUS_AUTO);
```

Comment:
I haven't seen this function before. What surprised me, is how the rest of the code is worked. I have taken a look for `update_auto_increment` usages, and I found that `ha_partition::write_row` does the following:
```
  if (have_auto_increment)
  {
    if (!table_share->next_number_keypart)
      update_next_auto_inc_val();
    error= update_auto_increment();
```
and `update_next_auto_inc_val` in turn is basically info() call, but with additional check:
```
void ha_partition::update_next_auto_inc_val()
{
  if (!part_share->auto_inc_initialized ||
      need_info_for_auto_inc())
    info(HA_STATUS_AUTO);
}
```
So the first assumption is, shouldn't we also add an additional check? Maybe the function is not very lightweight, and as far as i can see inside its ha_partition implementation, it is really so, at least a lock is involved.

I would like the rest of the code which works fine wouldn't be abused by another call, which it doesn't need.

So the suggestion is, since we have a partition only problem, to move the check somewhere to ha_partition as well.
I don't see an elegant way to do it, so another solution could be to leave this check in `TABLE::period_make_insert`: this feature is quite fresh and was never profiled yet, so this will not abuse most of the use cases.
Also, is this problem related only to SERIAL type? Then maybe in ideal world, should it be a virtual function in Field, which is overriden in the corresponding subclass, which is called from ha_partition-overriden function?